### PR TITLE
[SPARK-5995] [ML] Make Prediction dev API public

### PR DESCRIPTION
**Pull request by jkbradley**
Changes:
- Update protected prediction methods, following design doc. **<--most interesting change**
- Changed abstract classes for Estimator and Model to be public.  Added DeveloperApi tag.  (I kept the traits for Estimator/Model Params private.)
- Changed ProbabilisticClassificationModel method names to use probability instead of probabilities.

CC: @mengxr @shivaram @etrain

### CODE REVIEW

**Comment by shivaram** [comment]
Thanks @jkbradley - Change mostly looks good to me. I also notice you've changed the LogisticRegression to not have the custom transform ? Is that not required anymore ?

**Comment by jkbradley** [comment]
@shivaram  The updated ClassificationModel and ProbabilisticClassificationModel protected prediction methods *should* be more efficient now, so I don't think a specialized one for LogisticRegression is warranted.
Thanks for checking the PR!

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala [hunk 1]
```diff
@@ -150,6 +150,26 @@ sealed trait Vector extends Serializable {
       toDense
     }
   }
+
+  /**
+   * Find the index of a maximal element.  Returns the first maximal element in case of a tie.
+   * Returns -1 if vector has length 0.
+   */
+  private[spark] def findMax: Int = {
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala]: `findMax` -> `argmax`? I think we can put this in a separate PR or move it to `DenseVector` only. The problem is with sparse vectors, where we need to return an index with value zero if nonzero elements are negative.

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala [hunk 2]
```diff
@@ -150,6 +150,26 @@ sealed trait Vector extends Serializable {
...
+    if (size == 0) {
+      0
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala]: `-1`

We can also save `if .... else`:

~~~scala
var maxIdx = -1
var maxValue = Double.NegativeInfinity
foreachActive { (idx, value) =>
  ...
~~~

Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala]: unless it's a vector with values all NegativeInfinity
I'll fix the -1

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala (hunk 1 above)
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala]: Oh, good point.  I'll move it to DenseVector.  We've often added helper methods in random places, rather than to linear algebra, utils, or the place they really should belong.  I prefer we add these methods directly to the relevant place but make them private.

**Comment by jkbradley** [comment]
Fixed!  Any other comments?

**Comment by mengxr** [comment]
This looks good to me. There are two issues we may need to deal with in the future:

1. Where to put multi-label classifiers in the class hierarchy?
2. How to support distributively stored models? Those models could make predictions on individual records if the method call is triggered on the driver node. However, the `predict` method cannot be used in an RDD closure.

**Comment by mengxr** [comment]
Merged into master and branch-1.4. Thanks!

**Comment by jkbradley** [comment]
Made JIRAs for those 2 items:
* multilabel abstractions: [https://issues.apache.org/jira/browse/SPARK-7409]
* distributed models: [https://issues.apache.org/jira/browse/SPARK-7412]
